contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,876 @@
+        } catch Error(string memory reason) {
+            emit RewardDistributeFailed(operatorAddress, bytes(reason));
+        } catch (bytes memory lowLevelData) {
+            emit RewardDistributeFailed(operatorAddress, lowLevelData);
+        }
+    }
+
+    /**
+     * @dev Get new eligible validators from consensus engine
+     */
+    function updateEligibleValidators(
+        address[] calldata validators,
+        uint64[] calldata votingPowers
+    ) external onlyInitialized onlyCoinbase onlyZeroGasPrice {
+        uint256 newLength = validators.length;
+        if (newLength == 0) {
+            return;
+        }
+        uint256 oldLength = _eligibleValidators.length;
+        if (oldLength > newLength) {
+            for (uint256 i = newLength; i < oldLength; ++i) {
+                _eligibleValidators.pop();
+                _eligibleValidatorVoteAddrs.pop();
+            }
+        }
+
+        uint256 j;
+        for (uint256 i; i < newLength; ++i) {
+            address consensusAddress = validators[i];
+            address operatorAddress = _consensusToOperator[consensusAddress];
+            if (operatorAddress == address(0)) {
+                continue;
+            }
+
+            if (j >= oldLength) {
+                _eligibleValidators.push(
+                    IBSCValidatorSet.Validator({
+                        consensusAddress: consensusAddress,
+                        feeAddress: payable(0),
+                        BBCFeeAddress: address(0),
+                        votingPower: votingPowers[i],
+                        jailed: false,
+                        incoming: 0
+                    })
+                );
+                _eligibleValidatorVoteAddrs.push(_validators[operatorAddress].voteAddress);
+            } else {
+                IBSCValidatorSet.Validator storage eVal = _eligibleValidators[j];
+                eVal.consensusAddress = consensusAddress;
+                eVal.votingPower = votingPowers[i];
+                _eligibleValidatorVoteAddrs[j] = _validators[operatorAddress].voteAddress;
+            }
+            ++j;
+        }
+    }
+
+    function downtimeSlash(address consensusAddress, uint256 height) external onlyInitialized onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.poolModule != address(0), "VALIDATOR_NOT_EXIST");
+
+        // check slash record
+        bytes32 slashKey = _getSlashKey(operatorAddress, height, SlashType.DownTime);
+        SlashRecord storage record = _slashRecords[slashKey];
+        require(record.slashHeight == 0, "SLASHED");
+
+        // slash
+        uint256 slashAmount = IStakePool(valInfo.poolModule).slash(downtimeSlashAmount);
+        _jailValidator(valInfo);
```

Comment:
`_jailValidator` is unnecessary, because the consensus engine will do the sort.

Reply:
`_jailValidator` will do two things:
1. check and set `jailed` to true if needed; 2. remove from eligible validators.
You mean 2. is unnecessary right?

Reply:
Yes, I mean #2.

As I check the latest implementation of `getValidatorWithVotingPower`:
```
        for (uint256 i; i < count; ++i) {
            address operatorAddress = _validatorSet.at(offset + i);
            Validator memory valInfo = _validators[operatorAddress];
            consensusAddrs[i] = valInfo.consensusAddress;
            votingPowers[i] = valInfo.jailed ? 0 : IStakeCredit(valInfo.creditContract).totalPooledBNB();
        }
```

The bug is that: if there is only 5 validatos, and 2 are jailed, we need 21 validators, then all 5 validator will be choosen by consensus engine, but we only expect 3.

Reply:
No. The consensus engine will discard validator with zero voting powers. So in this case there will be only 3 eligible validators